### PR TITLE
settings: Drop experimental prefix from auto rehoming

### LINKED ISSUE #89921: MR: Rename experimental_enable_auto_rehoming cluster setting to drop the experimental prefix
With the auto rehoming feature sufficiently tested using the tpcc workload for multi region clusters, we can drop the experimental
designation from the cluster setting to enable auto rehoming.

Jira issue: CRDB-20502

Epic CRDB-16413

### PR DESCRIPTION
**Pull request by rimadeodhar**
The auto rehoming feature has been tested with the tpcc
workload and rehomes rows as expected based on the gateway region
of the originating UPDATE request. However, this feature leverages
the use of an ON UPDATE clause on the crdb_region column.
This precludes crdb_region from being used as a foreign key with the
ON UPDATE CASCADE option which is a common use case.
Until this shortcoming is fixed, we want to hide the cluster
setting and mark the session variable to enable auto rehoming
as a preview only feature.
However, in keeping with our updated nomenclature rules, we
still want to rename the session variable to drop the experimental prefix
We will keep the previous session variable name with the experimental
prefix as an alias to prevent errors on existing clusters that use
the session variable with the experimental prefix.
To summarize, this PR addresses the following two changes:
- Mark the cluster variable as hidden.
- Rename experimental_enable_auto_rehoming session variable to drop
  the experimental prefix. Create an alias named
  experimental_enable_auto_rehoming for the renamed
  enable_auto_rehoming session variable.

Resolves https://github.com/cockroachdb/cockroach/issues/89921

Release note (sql change): marked
sql.defaults.experimental_auto_rehoming.enabled cluster
setting as hidden. Also, renamed 'experimental_enable_auto_rehoming'
session variable to 'enable_auto_rehoming' and created an alias
'experimental_enable_auto_rehoming' for the renamed session variable.

### CODE REVIEW

**Review by rimadeodhar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89922)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm)

___
*[`pkg/sql/exec_util.go` line 303 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-NEHih3VC2rnSXmoatQ1:-NEHih3WERGSA726Vm62:b-76fus5) ([raw file](https://github.com/cockroachdb/cockroach/blob/de2360b1b60360dfd807f282baef7624a8f86717/pkg/sql/exec_util.go#L303)):*
> ```Go
> 	"default value for enable_auto_rehoming;"+
> 		" allows for rows in REGIONAL BY ROW tables to be auto-rehomed on UPDATE",
> 	false,
> ```

@ajstorm, @awoods187: would we want this cluster setting to be enabled by default?

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89922)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @awoods187, and @rimadeodhar)

___
*[`-- commits` line 14 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-NEHzYlWAJ8i-usxcXNN):*
nit: this should also mention the session variable name change
___
*[`pkg/settings/registry.go` line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-NEHwADM37Xu_CvA11n7:-NEHwADN6OIdGxzmo16G:bxk2jkj) ([raw file](https://github.com/cockroachdb/cockroach/blob/de2360b1b60360dfd807f282baef7624a8f86717/pkg/settings/registry.go#L147)):*
> ```Go
> 	"kv.refresh_range.time_bound_iterators.enabled":             {},
> 	"sql.defaults.datestyle.enabled":                            {},
> 	"sql.defaults.intervalstyle.enabled":                        {},
> ```

the old experimental cluster setting name should be added here, otherwise there will be lots of warning log spam for anyone who has set it
___
*[`pkg/sql/exec_util.go` line 303 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-NEHih3VC2rnSXmoatQ1:-NEHw-Dz8YVeGxJ8A0Uy:bsh759) ([raw file](https://github.com/cockroachdb/cockroach/blob/de2360b1b60360dfd807f282baef7624a8f86717/pkg/sql/exec_util.go#L303)):*
<details><summary><i>Previously, rimadeodhar (Rima Deodhar) wrote…</i></summary><blockquote>

@ajstorm, @awoods187: would we want this cluster setting to be enabled by default?
</blockquote></details>

if the feature is not experimental, then i would vote yes.

for anyone who has already set `sql.defaults.experimental_auto_rehoming.enabled=true`, then once they upgrade to the new version where the name is `sql.defaults.auto_rehoming.enabled` (and it defaults to false), their previous setting would be ignored.

at least if we change the default to true, we can communicate the name and default value change to the users at the same time
___
*[`pkg/sql/vars.go` line 1533 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-NEHvdR4BU5MWCmtE4ql:-NEHvdR4BU5MWCmtE4qm:b-4xz0ph) ([raw file](https://github.com/cockroachdb/cockroach/blob/de2360b1b60360dfd807f282baef7624a8f86717/pkg/sql/vars.go#L1533)):*
> ```Go
> 
> 	// CockroachDB extension.
> 	`enable_auto_rehoming`: {
> ```

instead of renaming, let's alias it so the old name works too. otherwise it would break any app that is referencing the variable.

take a look at how `idle_session_timeout` is defined below

**Review by rimadeodhar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89922)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @awoods187, and @rafiss)

___
*[`-- commits` line 14 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-NEHzYlWAJ8i-usxcXNN):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: this should also mention the session variable name change
</blockquote></details>

Done.
___
*[`pkg/settings/registry.go` line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-NEHwADM37Xu_CvA11n7:-NEIR8EP0rXsEDXh2oR0:b52ylnf) ([raw file](https://github.com/cockroachdb/cockroach/blob/de2360b1b60360dfd807f282baef7624a8f86717/pkg/settings/registry.go#L147)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

the old experimental cluster setting name should be added here, otherwise there will be lots of warning log spam for anyone who has set it
</blockquote></details>

Done, thanks for the callout! I didn't know this.
___
*[`pkg/sql/exec_util.go` line 303 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-NEHih3VC2rnSXmoatQ1:-NEIRDPv4MAFyF6u1186:b-nnrcmm) ([raw file](https://github.com/cockroachdb/cockroach/blob/de2360b1b60360dfd807f282baef7624a8f86717/pkg/sql/exec_util.go#L303)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

if the feature is not experimental, then i would vote yes.

for anyone who has already set `sql.defaults.experimental_auto_rehoming.enabled=true`, then once they upgrade to the new version where the name is `sql.defaults.auto_rehoming.enabled` (and it defaults to false), their previous setting would be ignored.

at least if we change the default to true, we can communicate the name and default value change to the users at the same time
</blockquote></details>

Sounds reasonable. Have enabled it to be true by default. @ajstorm, @awoods187 let me know if you feel differently.
___
*[`pkg/sql/vars.go` line 1533 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-NEHvdR4BU5MWCmtE4ql:-NEIRJoW54-KCBUzsA52:b9ucgam) ([raw file](https://github.com/cockroachdb/cockroach/blob/de2360b1b60360dfd807f282baef7624a8f86717/pkg/sql/vars.go#L1533)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

instead of renaming, let's alias it so the old name works too. otherwise it would break any app that is referencing the variable.

take a look at how `idle_session_timeout` is defined below
</blockquote></details>

Makes sense. Updated the code accordingly.

**Review by rafiss — CHANGES_REQUESTED** [review]
marking as reviewed; see previous comment

**Review by rafiss — CHANGES_REQUESTED** [review]
nice, this is a good way forwards! i left a small nit

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-:-NEfuORj2NnTkH7jOiUF:b1x00c7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @awoods187, and @rimadeodhar)

___
*[`pkg/sql/vars.go` line 1533 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-NEfu9_nE_J_24mLD7E3:-NEfu9_nE_J_24mLD7E4:b-9k9hp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2613a24bd2cf458e200b4ace59d7eb2b8d0aaccc/pkg/sql/vars.go#L1533)):*
> ```Go
> 
> 	// CockroachDB extension.
> 	`experimental_enable_auto_rehoming`: {
> ```

i think if we rename this (and the strings below) to `enable_auto_rehoming`, then the error message will be improved
___
*[`pkg/sql/vars.go` line 2362 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-NEfuHRu0drEjVM8wETR:-NEfuHRvC7YpXcKqjgTK:bbeht0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/2613a24bd2cf458e200b4ace59d7eb2b8d0aaccc/pkg/sql/vars.go#L2362)):*
> ```Go
> 	// We create `idle_in_session_timeout` before its existence.
> 	varGen[`idle_session_timeout`] = varGen[`idle_in_session_timeout`]
> 	varGen[`enable_auto_rehoming`] = varGen[`experimental_enable_auto_rehoming`]
> ```

and then also change this line to:
```
varGen[`experimental_enable_auto_rehoming`] = varGen[`enable_auto_rehoming`]
```
___
*[`pkg/sql/logictest/testdata/logic_test/set` line 854 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-NEftzIJ9-uTsfyymEK3:-NEftzIJ9-uTsfyymEK4:b95ubg0) ([raw file](https://github.com/cockroachdb/cockroach/blob/2613a24bd2cf458e200b4ace59d7eb2b8d0aaccc/pkg/sql/logictest/testdata/logic_test/set#L854)):*
> ```
> off
> 
> statement error pq: parameter "experimental_enable_auto_rehoming" requires a Boolean value
> ```

nit: we can probably change the variable definition so that the error message also no longer contains "experimental" in it. see my other comment

**Review by rimadeodhar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89922)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @awoods187, and @rafiss)

___
*[`pkg/sql/vars.go` line 1533 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-NEfu9_nE_J_24mLD7E3:-NEg9sd7BTv9m-gDSuPu:b-3oavw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/2613a24bd2cf458e200b4ace59d7eb2b8d0aaccc/pkg/sql/vars.go#L1533)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think if we rename this (and the strings below) to `enable_auto_rehoming`, then the error message will be improved
</blockquote></details>

Yes, you're right. Updated this.
___
*[`pkg/sql/vars.go` line 2362 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-NEfuHRu0drEjVM8wETR:-NEg9vuU5ZMnTlXBcFhU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2613a24bd2cf458e200b4ace59d7eb2b8d0aaccc/pkg/sql/vars.go#L2362)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

and then also change this line to:
```
varGen[`experimental_enable_auto_rehoming`] = varGen[`enable_auto_rehoming`]
```
</blockquote></details>

Done.

**Review by rafiss — APPROVED** [review]
lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89922#-:-NEgA2M3EwQmYZam4OLw:b-7nbn20)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @awoods187)

**Comment by rimadeodhar** [comment]
TFTR!

bors r=rafiss
